core/src/main/java/org/apache/calcite/rex/RexSimplify.java

```diff
@@ -676,52 +676,64 @@ private RexNode simplifyCoalesce(RexCall call) {
 
   private RexNode simplifyCase(RexCall call, RexUnknownAs unknownAs) {
     List<CaseBranch> inputBranches =
-        CaseBranch.fromCaseOperands(rexBuilder, new ArrayList(call.getOperands()));
+        CaseBranch.fromCaseOperands(rexBuilder, new ArrayList<>(call.getOperands()));
 
     // run simplification on all operands
-    RexSimplify branchSimplifier = this;
-    RelDataType branchType = call.getType();
+    RexSimplify condSimplifier = this.withPredicates(RelOptPredicateList.EMPTY);
+    RexSimplify valueSimplifier = this;
+    RelDataType caseType = call.getType();
 
+    boolean merged = false;
+    CaseBranch lastBranch = null;
     List<CaseBranch> branches = new ArrayList<>();
     for (CaseBranch branch : inputBranches) {
       // simplify the condition
-      RexNode newCond = branchSimplifier.simplify(branch.cond, RexUnknownAs.FALSE);
+      RexNode newCond = condSimplifier.simplify(branch.cond, RexUnknownAs.FALSE);
+      if (newCond.isAlwaysFalse()) {
+        // If the condition is false, we do not need to add it
+        continue;
+      }
 
-      // use the condition to simplify the branch
-      RexNode value = branch.value;
-      RexNode newValue = branchSimplifier.simplify(value, unknownAs);
+      // simplify the value
+      RexNode newValue = valueSimplifier.simplify(branch.value, unknownAs);
 
-      branches.add(new CaseBranch(newCond, newValue));
-    }
-
-    // remove branches with invalid conditions
-    branches.removeIf(branch -> branch.cond.isAlwaysFalse());
+      // create new branch
+      if (lastBranch != null) {
+        if (lastBranch.value.toString().equals(newValue.toString())
```

Comment:
I guess this never works since `newCond` is overwritten on each `for` iteration

Reply:
newCond becomes part of lastBranch, not sure if you are referring to that.

Reply:
That's truly hard to follow

Reply:
This lets us simplifying the merged condition only once. I will add some additional comments with explanation so it becomes a bit easier to know what is going on.